### ISSUE TITLE
MPS: mul of an MPS tensor by a CPU scalar tensor binds a read-only buffer to a write-enabled Metal argument (mul_dense_scalar_float_float validation assertion   / GPU fault)

### ISSUE THREAD

**Issue by Bill-Saccadic-AI** [pytorch/pytorch#186580]
### 🐛 Describe the bug

  Multiplying an MPS tensor by a **CPU 0-dim scalar tensor** drives the MPS
  `mul_dense_scalar_float_float` compute kernel to bind the read-only scalar
  bytes to a Metal shader argument that has **write access enabled**. With Metal
  API validation on this is a hard assertion; in real workloads (the same op
  inside a forward pass with other device-mixed arithmetic) it surfaces as an
  uncatchable GPU command-buffer fault — the process dies with SIGTRAP/abort and
  no Python traceback.

  Mixing an MPS tensor with a CPU scalar tensor is a normally-supported pattern
  (it works on CPU and CUDA, which promote the scalar), so this is a backend
  correctness/safety bug rather than user error.

  ### To reproduce
  
  ```python
  import torch
  m = torch.ones(4, 4, device="mps")   # any MPS float tensor
  s = torch.tensor(2.0)                 # a CPU 0-dim scalar tensor
  out = m * s                           # -> aten::mul_dense_scalar_float_float
  torch.mps.synchronize()
  print(out.sum())
  
  Run with Metal validation enabled to make it deterministic:
  
  MTL_DEBUG_LAYER=1 MTL_DEBUG_LAYER_ERROR_MODE=assert python repro.py

  Output:

  validateComputeFunctionArguments:1066: failed assertion
  `Compute Function(mul_dense_scalar_float_float): Read-only bytes are being
   bound at index 2 to a shader argument with write access enabled
   (did you mean to use const or constant in the shader?).'

  Note: without MTL_DEBUG_LAYER=1 the tiny repro above may appear to succeed
  (the malformed binding is tolerated when the kernel doesn't actually write the
  slot), but the validation layer flags it deterministically, and in larger
  graphs the same *_dense_scalar_* device-mixed path produces a real GPU fault
  (observed as SIGTRAP, rc=133, mid-forward). A closely-related sub of an MPS
  tensor by a CPU scalar (sub_strided_cast_float : CPU:Long[] : MPS:Float[...])
  appears on the same code path.

  Expected behavior

  Parity with the CPU/CUDA backends: the CPU scalar should be promoted to the
  device (or bound as a read-only/constant Metal argument). No Metal argument
  binding violation, and no GPU fault.

  Workaround

  Use a Python float, or move the scalar to MPS first:

  out = m * 2.0          # ok
  out = m * s.to("mps")  # ok

  The assertion's own hint ("did you mean to use const or constant in the
  shader?") points at the fix: the scalar argument in the *_dense_scalar_*
  Metal shader should be declared read-only/constant.

  Versions

  - PyTorch: 2.11.0 — also reproduces on 2.12.0 (tested)
  - OS: macOS 26.5 (build 25F71), arm64
  - Device: Apple M4 (16 GB unified memory)
  - Python: 3.12.13

### Versions

  PyTorch version: 2.11.0   (also reproduces on 2.12.0)
  Is debug build: False
  CUDA used to build PyTorch: None
  ROCM used to build PyTorch: N/A

  OS: macOS 26.5 (arm64)
  Clang version: 21.0.0 (clang-2100.1.1.101)
  Python version: 3.12.13 (main, Mar  3 2026, 12:39:30) [Clang 17.0.0 (clang-1700.6.3.2)] (64-bit runtime)
  Python platform: macOS-26.5-arm64-arm-64bit
  Is CUDA available: False
  Is XPU available: False
  Is XNNPACK available: True

  CPU:
  Apple M4

  Versions of relevant libraries:
  [pip3] numpy==2.4.4
  [pip3] torch==2.11.0

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

**Comment by Bill-Saccadic-AI** [comment]
On re-test the *_dense_scalar assertion only appears under MTL_DEBUG_LAYER=1 and has no functional impact; closing."
